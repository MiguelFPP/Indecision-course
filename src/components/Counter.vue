<template>
    <h2>{{ customTitle }}</h2>
    <p>{{ counter }}</p>
    <p data-testid="counter">{{ counter }}</p>

    <button v-on:click="increment">+</button>
    <button v-on:click="decrement">-</button>

    <p v-if="message">{{ message }}</p>

    <h2>Square value</h2>

    <p>{{ this.start }} <sup>2</sup> = {{ getSquaredCounted }} </p>
</template>

<script>
export default {
    name: 'Counter',
    props: {
        title: { type: String, required: true },
        start: {
            type: Number, default: 100,
            validator(value) {
                return value >= 100;
            }
        },
    },
    data() {
        return {
            counter: this.start,
            message: null
        }
    },
    methods: {
        increment() {
            this.counter++;

            if (this.message) {
                this.message = null;
            }
        },
        decrement() {
            if (this.counter > 0) {
                this.counter--;
            } else {
                this.message = "Counter can't be less than 0";
            }
        },
        getSquaredValue() {
            return this.counter * this.counter;
        },
    },
    computed: { //se lanza una vez que se ha cambiado el valor de counter
        getSquaredCounted() {
            console.log('getSquaredCounted');
            return this.counter * this.counter;
        },
        customTitle() {
            return this.title ? this.title : 'Counter';
        }
    },

}
</script>

<style>
button {
    background-color: #4CAF50;
    border-radius: 5px;
    border: none;
    color: white;
    margin: 0 5px;
    padding: 10px 20px;
    text-align: center;
    text-decoration: none;
    cursor: pointer;
}

button:hover {
    background-color: #3e8e41;
}
</style>