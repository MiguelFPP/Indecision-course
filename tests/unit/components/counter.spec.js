import { shallowMount, mount } from "@vue/test-utils"; // para montar los componentes
import Counter from "@/components/Counter";

describe('Counter component', () => {
    let wrapper;
    beforeEach(() => {
        return wrapper = shallowMount(Counter);
    });
    /* test('debe hacer match con el snapshot', () => {
        const wrapper = shallowMount(Counter);

        expect(wrapper.html()).toMatchSnapshot()
    }); */

    //evaluar el valor de las etiques
    test('Debe de tener el valor por defecto Counter', () => {

        const tag = wrapper.find('h2');

        expect(tag.exists()).toBeTruthy();
        expect(tag.text()).toBe('Counter');
    })

    test('El valor por defecto ser 100 el p', () => {

        const p_tags = wrapper.find('[data-testid="counter"]');

        expect(p_tags.exists()).toBeTruthy();
        expect(p_tags.text()).toBe('100');
    })

    test('Debe de incrementar y decrementar el valor de counter', async () => { //Hacer uso de async y await para usar evento en pruebas

        const button_tags = wrapper.findAll('button');

        expect(button_tags[0].exists()).toBeTruthy()
        await button_tags[0].trigger('click');

        let p_tags = wrapper.find('[data-testid="counter"]');
        expect(p_tags.text()).toBe('101');

        await button_tags[1].trigger('click');
        await button_tags[1].trigger('click');

        p_tags = wrapper.find('[data-testid="counter"]');
        expect(p_tags.text()).toBe('99');
    });

    test('debe ded establecer el valor por defecto', () => {
        const { start } = wrapper.props(); //definido de tipo Number

        const value = wrapper.find('[data-testid="counter"]').text();//llega como objeto string

        expect(Number(value)).toBe(start);
    })

    test('Debe de mostrar la prop title', () => {
        const title = 'Hola mundo';
        const start_number = 100;
        const wrapper = shallowMount(Counter, {
            props: {
                title,
                start: start_number
            }
        });

        expect(wrapper.find('h2').text()).toBe(title);

        expect(Number(wrapper.find('[data-testid="counter"]').text())).toBe(start_number);// type number
    });
})