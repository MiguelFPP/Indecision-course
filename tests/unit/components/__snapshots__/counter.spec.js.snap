// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component debe hacer match con el snapshot 1`] = `
<h2>Counter</h2>
<p>1</p>
<button>+</button>
<button>-</button>
<!--v-if-->
<h2>Square value</h2>
<p>1 <sup>2</sup> = 1</p>
`;
